src/core/server/metrics/collectors/cgroup.ts

```diff
@@ -0,0 +1,177 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import fs from 'fs';
+import { join as joinPath } from 'path';
+import { MetricsCollector, OpsOsMetrics } from './types';
+
+type OsCgroupMetrics = Pick<OpsOsMetrics, 'cpu' | 'cpuacct'>;
+
+interface OsCgroupMetricsCollectorOptions {
+  cpuPath?: string;
+  cpuAcctPath?: string;
+}
+
+export class OsCgroupMetricsCollector implements MetricsCollector<OsCgroupMetrics> {
+  // Used to prevent unnecessary file reads on systems not using cgroups
+  private noCgroupPresent = false;
+
+  constructor(private readonly options: OsCgroupMetricsCollectorOptions) {}
+
+  public async collect(): Promise<OsCgroupMetrics> {
+    if (this.noCgroupPresent) {
+      return {};
+    }
+
+    try {
+      const cgroups = await readControlGroups();
+      const cpuPath = this.options.cpuPath || cgroups[GROUP_CPU];
+      const cpuAcctPath = this.options.cpuAcctPath || cgroups[GROUP_CPUACCT];
```

Comment:
Maybe `readControlGroups` should only be called if `this.options.cpuPath` and `this.options.cpuAcctPath` are not defined? Also the `cgroups` value can probably be memoized, to avoid accessing / reading the file at every `collect` call.